### PR TITLE
Sockets - Kate N

### PR DESCRIPTION
**Pull request by KateAnnNichols**
# JS Adagrams
Congratulations! You're submitting your assignment!
## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? |  loops, hashes, arrays 
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | Not in particular- It seems like JS resources and tutorials are actually easier to find 
What was a challenge you were able to overcome on this assignment? | Sorting out logic for breaking a tie
What has been interesting and positive about learning a new programming language? | Learning how important JS is for the web 
What is something to focus on your learnings in JavaScript in the next week? | TDD in JS

### CODE REVIEW

File: src/adagrams.js [hunk 1]
```diff
@@ -1,8 +1,133 @@
...
+// WAVE 1 -->
+// use entries 
+// creating the empty lettersPool array 
+    const lettersPool = [];
+// loop through frequencies to push letters to lettersPool 
+// entries will be an array of arrays w/ letter + count in each array 
+    const entries = Object.entries(letterFrequencies);
+    entries.forEach(entry => {
+      letter = entry[0];
+      count = entry[1];
```
Comment (droberts-sea) [inline src/adagrams.js]: Two things:

1. Using `Object.entries` here gets the job done, but it's not super readable code. I might do something with `Object.keys` instead:
    ```js
    Object.keys(letterFrequencies).forEach(letter => {
      const count = letterFrequencies[letter];
      // ...
    ```

2. You're defining `letter` and `count` as global variables here! You should prefix them with `const` (why not `let`?)

File: src/adagrams.js [hunk 2]
```diff
@@ -1,8 +1,133 @@
...
+// copying lettersInHand array so that it can be changed w/o changing original array 
+    let letters = lettersInHand;
+// initialization - condition - increment 
+    for (let i = 0; i < input.length; i += 1) {
+// check for input and use indexOf method (tells index in the string where letter is- if letter not found- returns -1 so will be false) 
+// splice method (add/subtract - permanently changes array - to prevent false positive)
+// return true if every letter in the input word is available in correct quantity from lettersInHand 
+      let letter = letters.indexOf(input[i]);
+      if (letter === -1) {
+        return false;
```
Comment (droberts-sea) [inline src/adagrams.js:31]: You call the variable on line 57 `letter`, but it doesn't hold a letter, it holds an index. Could you choose a better name for this variable? That would make the `splice` operation on line 61 a little more clear.

File: src/adagrams.js [hunk 3]
```diff
@@ -1,8 +1,133 @@
...
+// check function
+
+// WAVE 2: check if the word only uses some/all letters from handLetters array 
+// method (usesAvailableLetters w/ 2 parameters (input + lettersInHand))
+// input is a string for input word from player
+// lettersInHand is an array of 10 strings - each string is a letter 
+  usesAvailableLetters(input, lettersInHand) {
+// copying lettersInHand array so that it can be changed w/o changing original array 
+    let letters = lettersInHand;
+// initialization - condition - increment
```
Comment (droberts-sea) [inline src/adagrams.js]: Line 51 does not copy the array itself, it creates a new variable pointing at the same array. If you want a copy, you would need to do something like `const letters = [...lettersInHand];`.

File: src/adagrams.js [hunk 4]
```diff
@@ -1,8 +1,133 @@
...
+    if (word.length >= 7 && word.length < 11) {
+      score = 8;
+    }; 
+// make the word into an array of strings w/ each letter as a string 
+// use forEach like in tips.js - more concise than just for 
+// use arrow function to minimize syntax in forEach - make 1 line instead of 3
+    word.split('').forEach(ltr => {
+// [ltr] for key in hash (score value)
+      score += letterValues[ltr];
+    });
```
Comment (droberts-sea) [inline src/adagrams.js:56]: You should spell `letter` out fully here.

File: src/adagrams.js [hunk 5]
```diff
@@ -1,8 +1,87 @@
...
+      if (score > highScore) {
+        highScore = score;
+        word = words[i];
+      } else if (score == highScore) {
+      word = this.breakTie(word, words[i]);
+    }
+  }
+  return {word: word, score: highScore}; 
+},
 };
```
Comment (droberts-sea) [inline src/adagrams.js:85]: Watch your indentation through here.

File: src/adagrams.js [hunk 6]
```diff
@@ -1,8 +1,87 @@
...
+      }
+      score += letterValues[letter];
+    });
+    return score; 
+  },
+
+  breakTie(first, second) {
+      if (first.length == second.length) {
+        return first;
+      } else if (first.length == 10) {
```
Comment (droberts-sea) [inline src/adagrams.js:63]: I love that you chose to break this tie-breaking logic out as a separate method - this helps make the code organized and easy to read. Good functional decomposition!

**Comment by droberts-sea** [comment]
# JS Adagrams
## What We're Looking For



Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Small commits with meaningful commit messages | There is still a lot of room for improvement here. Your commit messages should tell the reader a story about what changed and why, without them having to refer to the project spec. Messages like "Wave 2" or "Updated adagrams.js" do not accomplish this.<br><br>Also, it looks like you're still using the GitHub web interface to update files. At this point you should be doing all your git work through the terminal.
**Code Requirements** |
`drawLetters` method | yes
Uses appropriate data structure to store the letter distribution | yes - I like the choice to store frequencies as numbers instead of writing a big array full of repeated letters by hand.
All tests for `drawLetters` pass | yes
`usesAvailableLetters` method | yes
All tests for `usesAvailableLetters` pass | yes
`scoreWord` method | yes
Uses appropriate data structure to store the letter scores | yes
All tests for `scoreWord` pass |
`highestScoreFrom` method | yes
Appropriately handles edge cases for tie-breaking logic | yes
All tests for `highestScoreFrom` pass | yes
**Overall** | Good job overall. This code is well-organized and easy to read, and you've made good choices about what data structures to use to solve the different parts of it. I still see a lot of room for improvement around your git hygiene, but in general I'm happy with what I see. This submission does qualify as "green", meeting the requirements for this project that we established last week. Keep up the hard work!
